File: src/main/scala/com/kirkwoodwest/scalawoods/GridButton.scala
package com.kirkwoodwest.scalawoods

import com.bitwig.extension.controller.api.{HardwareButton, MidiIn, MidiOut}
import com.kirkwoodwest.scalawoods.GridButtonState

class GridButton(val button: HardwareButton, val lightData: LightData):
  protected var isDirty:Boolean = false
  protected var state: GridButtonState = GridButtonState.Off

  def setupNoteButton(midiIn: MidiIn, note: Int): Unit =
    button.pressedAction().setActionMatcher(midiIn.createNoteOnActionMatcher(0, note))
    button.releasedAction().setActionMatcher(midiIn.createNoteOffActionMatcher(0, note))

  def setupCCButton(midiIn: MidiIn, cc: Int): Unit =
    button.pressedAction().setActionMatcher(midiIn.createCCActionMatcher(0, cc, 127))
    button.releasedAction().setActionMatcher(midiIn.createCCActionMatcher(0, cc, 0))

  def setState(state: GridButtonState): Unit =
    this.state = state
    isDirty = true

  def setDirty(): Unit =
    isDirty = true

  def flush(): Unit =
    if(isDirty)
      isDirty = false
      var channel: Int = 0
      
      if (state.equals(GridButtonState.ClipPlaying) || state.equals(GridButtonState.ClipPlayingNote))
        channel = 2
      
      lightData.midiOut.sendMidi(0x90 +channel, lightData.note, this.state.value)

end GridButton

File: src/main/java/com/kirkwoodwest/launchpadminimk3/hardware/GridButton.java
package com.kirkwoodwest.launchpadminimk3.hardware;

import com.bitwig.extension.controller.api.HardwareButton;
import com.bitwig.extension.controller.api.MidiIn;

public class GridButton {
    private final HardwareButton button;
    private final LightData lightData;
    private boolean isDirty = false;
    private GridButtonState state = GridButtonState.Off;

    public GridButton(HardwareButton button, LightData lightData) {
        this.button = button;
        this.lightData = lightData;
    }

    public void setupNoteButton(MidiIn midiIn, int note) {
        button.pressedAction().setActionMatcher(midiIn.createNoteOnActionMatcher(0, note));
        button.releasedAction().setActionMatcher(midiIn.createNoteOffActionMatcher(0, note));
    }

    public void setupCCButton(MidiIn midiIn, int cc) {
        button.pressedAction().setActionMatcher(midiIn.createCCActionMatcher(0, cc, 127));
        button.releasedAction().setActionMatcher(midiIn.createCCActionMatcher(0, cc, 0));
    }

    public void setState(GridButtonState state) {
        this.state = state;
        isDirty = true;
    }

    public void setDirty() {
        isDirty = true;
    }

    public void flush() {
        if (isDirty) {
            isDirty = false;
            int channel = 0;

            if (state.equals(GridButtonState.ClipPlaying) || state.equals(GridButtonState.ClipPlayingNote)) {
                channel = 2;
            }

            lightData.getMidiOut().sendMidi(0x90 + channel, lightData.getNote(), this.state.getValue());
        }
    }

    public HardwareButton getButton() {
        return button;
    }

    public LightData getLightData() {
        return lightData;
    }
}